mlir-sycl/lib/Transforms/Inliner.cpp

```diff
@@ -172,47 +283,219 @@ class Inliner : public InlinerInterface {
+  auto WalkFn = [&](Operation *SymbolTableOp, bool AllUsesVisible) {
+    for (Operation &Op : SymbolTableOp->getRegion(0).getOps()) {
+      // If this is a callgraph operation, check to see if it is discardable.
+      if (auto Callable = dyn_cast<CallableOpInterface>(&Op)) {
+        if (auto *Node = CG.lookupNode(Callable.getCallableRegion())) {
+          SymbolOpInterface Symbol = dyn_cast<SymbolOpInterface>(&Op);
+          if (Symbol && (AllUsesVisible || Symbol.isPrivate()) &&
+              Symbol.canDiscardOnUseEmpty())
+            DiscardableSymNodeUses.try_emplace(Node, 0);
+          continue;
+        }
+      }
+      // Otherwise, check for any referenced nodes. These will be always-live.
+      walkReferencedSymbolNodes(&Op, CG, SymbolTable, AlwaysLiveNodes,
+                                [](CallGraphNode *, Operation *) {});
+    }
+  };
+
+  SymbolTable::walkSymbolTables(Op, /*allSymUsesVisible=*/!Op->getBlock(),
+                                WalkFn);
+
+  // Drop the use information for any discardable nodes that are always live.
+  for (auto &It : AlwaysLiveNodes)
+    DiscardableSymNodeUses.erase(It.second);
+
+  // Compute the uses for each of the callable nodes in the graph.
+  for (CallGraphNode *CGN : CG)
+    recomputeUses(CGN, CG);
+}
+
+void CGUseList::walkReferencedSymbolNodes(
+    Operation *Op, CallGraph &CG, SymbolTableCollection &SymbolTable,
+    DenseMap<Attribute, CallGraphNode *> &ResolvedRefs,
+    function_ref<void(CallGraphNode *, Operation *)> Callback) {
+  auto SymbolUses = SymbolTable::getSymbolUses(Op);
+  assert(SymbolUses && "expected uses to be valid");
+
+  Operation *SymbolTableOp = Op->getParentOp();
+  for (const SymbolTable::SymbolUse &Use : *SymbolUses) {
+    auto RefIt = ResolvedRefs.insert({Use.getSymbolRef(), nullptr});
+    CallGraphNode *&CGN = RefIt.first->second;
+
+    // If this is the first instance of this reference, try to resolve a
+    // callgraph node for it.
+    if (RefIt.second) {
+      auto *SymbolOp = SymbolTable.lookupNearestSymbolFrom(SymbolTableOp,
+                                                           Use.getSymbolRef());
+      auto CallableOp = dyn_cast_or_null<CallableOpInterface>(SymbolOp);
+      if (!CallableOp)
+        continue;
+      CGN = CG.lookupNode(CallableOp.getCallableRegion());
+    }
+    if (CGN)
+      Callback(CGN, Use.getUser());
+  }
+}
+
+bool CGUseList::isDead(CallGraphNode *CGN) const {
+  // If the parent operation isn't a symbol, simply check normal SSA deadness.
+  Operation *Op = CGN->getCallableRegion()->getParentOp();
+  if (!isa<SymbolOpInterface>(Op))
+    return isMemoryEffectFree(Op) && Op->use_empty();
+
+  // Otherwise, check the number of symbol uses.
+  auto SymbolIt = DiscardableSymNodeUses.find(CGN);
+  return SymbolIt != DiscardableSymNodeUses.end() && SymbolIt->second == 0;
+}
+
+bool CGUseList::hasOneUse(CallGraphNode *CGN) const {
```

Comment:
I think the name in Inliner.cpp is better:
```suggestion
bool CGUseList::hasOneUseAndDiscardable(CallGraphNode *CGN) const {
```